Add multiple quantity consumables checkout

veenone opened the pull request:
# Description

This branch extends checkout functionality for consumable items to support multiple number of item per check out.
Each checkout will have its own description for historical purpose.
Remaining item calculation is now based on total item checked out

Fixes #1820 #5176

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Test A : validating all functions working as intended and mitigate implementation side effects (other view using the updated data. i.e : consumables view in user)
- [x] Test B : validating on error handler during quantity comparison between total stock, remaining and checked out stock

**Test Configuration**:
* PHP version: 7.4.27
* MySQL version : 10.4.22
* Webserver version: Apache 2.4.52
* OS version: Windows 10


# Checklist:

- [x] I have read the Contributing documentation available here: https://snipe-it.readme.io/docs/contributing-overview
- [x] I have formatted this PR according to the project guidelines: https://snipe-it.readme.io/docs/contributing-overview#pull-request-guidelines
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# screens

## consumable checkout
![checkout-consumables](https://user-images.githubusercontent.com/3839381/159167453-7f2464d2-ea60-4c7c-a454-b728cd792034.PNG)

## consumable details
![consumable-details](https://user-images.githubusercontent.com/3839381/159167497-1c3f9056-1e60-4bec-a8b4-cc7ba7ecf7c7.PNG)

## consumable stock validation
![consumable-stock checking](https://user-images.githubusercontent.com/3839381/159167509-a48892fa-a806-467d-abe5-e544f1bade67.PNG)

## user's consumable view
![user-consumable](https://user-images.githubusercontent.com/3839381/159167515-65562776-d91e-4e74-9f7f-c785d5d9f5a4.PNG)



# Remarks
I'm sorry I don't have a knowledge on web programming let alone PHP & laravel mvc implementation.
I'm open for any suggestion & guidance 🙇‍♂️

achmarah committed:
- modifying consumables checkout to support multiple amount
- adding total checkedout amount in consumables_user table
- redirecting note during checkout to checkoutnote in consumables_user table

veenone committed:
- fix wrong referencing of commit source
- add migrate file related to multiple consumables

veenone committed:
add missed mapping

veenone committed:
adding validation on consumable checkout

- adding validation on amount input compared to remaining amount
- change link navigation for cancel button

veenone committed:
fix pivot query of consumable in user page

- fix pivot query of consumable in user page
- update consumable information details in user page

snipe commented on app/Http/Controllers/Consumables/ConsumableCheckoutController.php:
```diff
@@ -48,6 +49,14 @@ public function store(Request $request, $consumableId)
         }
 
+        $request->validate([
+            "totalnum" => "required|regex:/^[0-9]*$/|max:{{$consumable->numRemaining()}}"
```
Blade syntax is only for blades - I don’t think this will work here

snipe replied:
(Even if it did, we’d want a custom validator here, so we could localize the strings)

veenone replied:
any advice you might have for the proper validation syntax location?

veenone replied:
string will be updated to support localization

veenone replied:
for now,  strings on validator has been moved to admin/consumables/message.
so far checking for these 3 validations work. but whenever I switch my locale to different language,  it doesn't switch. strange..

snipe commented on database/migrations/2013_11_15_190357_create_licenses_table.php:
```diff
@@ -0,0 +1,42 @@
+     *
+     * @return void
+     */
+    public function up()
```
Why is this here? This is a migration from 2013

veenone replied:
this one is a mistake from me. I mistakenly added the file because I didn't fast forward the repo after initial forking of the repo.

To be deleted

veenone committed:
- updating validator to use localized strings
- adding new localized string for validation on some languages

veenone committed:
remove 2013_11_15_190357_create_licenses_table.php which mistakenly added in the repo

achmarah committed:
updating amount validation logic during checkout

veenone committed:
incorporating number of checked out consumables in the email notification message

uberbrady commented on app/Events/CheckoutableCheckedOut.php:
```diff
@@ -14,17 +14,24 @@ class CheckoutableCheckedOut
     public $checkedOutTo;
     public $checkedOutBy;
     public $note;
+    public $total;
```
The attribute seems to be called `$totalnum` in some places, and `$total` here. Also I think we would have to add the `checkoutnote` field as well?

veenone replied:
ok I will review this

uberbrady commented on app/Models/Consumable.php:
```diff
@@ -285,13 +286,87 @@ public function getEula()
         $checkedout = $this->users->count();
+        foreach($this->users as $data){
+                $totalnum = $data['pivot']['totalnum'];
+                Log:info($totalnum);
```
This should probably be `Log::debug()` (two colons, and debug level).

veenone replied:
ah.. typo.. I will update it

uberbrady commented on app/Models/Consumable.php:
```diff
@@ -285,13 +286,87 @@ public function getEula()
+        $checkedouttotal = null;
         $checkedout = $this->users->count();
+        foreach($this->users as $data){
+                $totalnum = $data['pivot']['totalnum'];
```
This may work as-is, but we prefer to refer to this as: `$data->pivot->totalnum`

veenone replied:
to be updated

uberbrady commented on app/Models/Consumable.php:
```diff
@@ -285,13 +286,87 @@ public function getEula()
     {
+        $checkedouttotal = null;
         $checkedout = $this->users->count();
+        foreach($this->users as $data){
```
Usually variable names like `$data` are a bad idea - since this is a list of users, how about just `$user`?

veenone replied:
noted

uberbrady commented on app/Models/Consumable.php:
```diff
@@ -285,13 +286,87 @@ public function getEula()
+     * @since [v.6]
+     * @return int
+     */
+    public function numConsumed()
```
This looks a bit like copy-paste code from `numRemaining`, above. Can we do something either there or here that does `$total - $this->numConsumed()` so we don't duplicate the logic?

veenone replied:
this related to the usage checkoutConsumableNotification.php .  To be reviewed.

uberbrady commented on app/Models/Consumable.php:
```diff
@@ -285,13 +286,87 @@ public function getEula()
+    {
+        $checkedouttotal = null;      
+        $checkoutnote = $this->users()->orderBy('created_at','desc')->get();
+        $last = $checkoutnote->last()['pivot']['totalnum'];  
```
Same note about the `pivot` here - also can we change the name on this function? Something a little more clear like `lastConsumedQty()` or something?

veenone replied:
Noted :)

uberbrady commented on app/Models/Consumable.php:
```diff
@@ -285,13 +286,87 @@ public function getEula()
+     */
+    public function checkoutTotal()
+    {
+        $checkedout = $this->users->count();
```
Doesn't this ignore the `$total` or `$totalnum` or whatever it is we're calling it?

veenone replied:
I will review back with the naming of these total related variables

uberbrady commented on app/Models/Consumable.php:
```diff
@@ -285,13 +286,87 @@ public function getEula()
+     * @since [v.6]
+     * @return string
+     */
+    public function checkoutNote()
```
Similar to above, a better method name here would be very helpful - `lastCheckoutNote()` or something like that?

veenone replied:
Noted

uberbrady commented on app/Notifications/CheckoutConsumableNotification.php:
```diff
@@ -123,6 +123,8 @@ public function toMail()
                 'req_accept'    => $req_accept,
                 'accept_url'    => $accept_url,
+                'totalnum'      => $this->item->lastConsumed(),
+                'checkoutnote'  => $this->item->checkoutNote()
```
Ah! Now I see why you're doing the `lastConsumed()` and `checkoutNote()` methods above. Unfortunately, this could be subject to a race condition, in that if you did two checkouts of consumables-with-quantities at the exact same time, the wrong quantity and note might get picked up.

A better approach might be to instead fire the CheckoutConsumableNotification against the `users_consumables` record instead of specifying an Admin, and a Consumable, and a recipient. Or to at least include the quantity and checkoutnote in the constructor?

veenone replied:
I'm still wondering on the logic of parsing the information after checout to the email actually (pardon my limited knowledge on this topic 😅 ), is there any examples somewhere covering the same topic?

uberbrady commented on app/Services/PredefinedKitCheckoutService.php:
```diff
@@ -172,8 +172,10 @@ function () use ($user, $admin, $checkout_at, $expected_checkin, $errors, $asset
                         'consumable_id' => $consumable->id,
                         'user_id' => $admin->id,
                         'assigned_to' => $user->id,
+                        'totalnum' => $consumable->pivot->$totalnum,
```
Are these `$` supposed to be here?

veenone replied:
to be fixed

uberbrady commented on database/migrations/2013_11_25_013244_create_licenses_table.php:
```diff
@@ -0,0 +1,43 @@
+<?php
```
This file doesn't look like it belongs here.

veenone replied:
to be removed. I forgot to discard from my local repo

mikeroq replied:
Seems to still be here. I had to remove one of these to get migrations to run due to duplicate class names.

uberbrady commented on public/web.config:
```diff
@@ -23,5 +23,12 @@
         <remove fileExtension=".woff2" />
         <mimeMap fileExtension=".woff2" mimeType="application/x-font-woff2" />
     </staticContent>
+        <directoryBrowse enabled="true" />
```
I don't think we want this part of the change actually

veenone replied:
maybe this change due to my local environment setup.. I will revert back to the initial condition

uberbrady commented on resources/lang/en-GB/admin/consumables/general.php:
```diff
@@ -8,4 +8,5 @@
     'remaining' 			            => 'Remaining',
     'total' 			                => 'Total',
     'update'                            => 'Update Consumable',
+    'note'                              => 'Remarks',
```
So these additions to the language files need to happen in `en`, so they then get added to CrowdIn, and then you can add the new translations to CrowdIn. See the docs here: https://snipe-it.readme.io/docs/translations

veenone replied:
noted. next time I will add only in en. do you think I should remove the one I made already in the other languages?

uberbrady commented on resources/lang/en/admin/consumables/message.php:
```diff
@@ -4,32 +4,40 @@
     'does_not_exist' => 'Consumable does not exist.',
 
-    'create' => [
+    'create' => array(
```
If there's any way we can not include these unrelated changes, I would definitely appreciate it - makes it much easier to review.

veenone replied:
noted, to be reverted

veenone replied:
hmm strange.. after checking, mine is different with the one in develop projects.. maybe due to my initial mistake when first forking the project without fast forward. Fixed now

uberbrady commented on resources/views/consumables/checkout.blade.php:
```diff
@@ -21,7 +21,7 @@
           <div class="box-header with-border">
             <div class="box-heading">
-              <h2 class="box-title">{{ $consumable->name }} </h2>
+              <h2 class="box-title">{{ $consumable->name }}   ({{ $consumable->numRemaining()  }} out of {{ $consumable->qty}}  {{ trans('admin/consumables/general.remaining') }}) </h2>
```
Unfortunately you'll have to translate `out of` as well. You might be able to cheat by using a symbol such as `/` ?

e.g. instead of "3 out of 5" something like "3 / 5 remaining"? Not sure that will look right though...

veenone replied:
noted , proper update to be considered

uberbrady commented on resources/views/consumables/checkout.blade.php:
```diff
@@ -67,16 +81,16 @@
               </div>
             @endif
           <!-- Note -->
-          <div class="form-group {{ $errors->has('note') ? 'error' : '' }}">
```
So we *do* have a `notes` field in consumables (it's not actually `note`, but that's not your bug!) We still probably want to display that note for the consumable - imagine it's for toner and the `notes` field might say "This toner is for Canon 12345 and 234567 laser-printers, and the 5xg-LX super-duper laser-printer" - that still needs to show up.

So I don't think we can just 'steal' the field for note here.

veenone replied:
got it, note to be displayed in conjunction the checkout note

veenone replied:
the view is updated as below.
Logically, consumable note is not updateable, so I put the textarea to disable editing
![image](https://user-images.githubusercontent.com/3839381/162199328-d05d1264-d2b6-4299-95ae-3f796456289e.png)

uberbrady commented on resources/views/consumables/checkout.blade.php:
```diff
@@ -67,16 +81,16 @@
         </div> <!-- .box-body -->
         <div class="box-footer">
-          <a class="btn btn-link" href="{{ URL::previous() }}">{{ trans('button.cancel') }}</a>
+          <a class="btn btn-link" href="{{ route('consumables.index') }}">{{ trans('button.cancel') }}</a>
```
This is _probably_ fine? But If we don't need to change this I'd rather not, there's enough other stuff already going on...

veenone replied:
I updated this since I had a lot of checkouts and wanted to go back to the index by pressing back. but it keeps coming back to previous checkout processes that I did and it doesn't make any sense in my opinion.  maybe I'm missing the idea of the usage of the link, or I was abusing the link during my test.

To revert the implementation on this

veenone replied:
link reverted back in the last commit

uberbrady commented on resources/views/consumables/view.blade.php:
```diff
@@ -22,7 +22,7 @@
       <div class="box-header with-border">
         <div class="box-heading">
-          <h2 class="box-title"> {{ $consumable->name }}</h2>
+          <h2 class="box-title"> {{ $consumable->name }}   ({{ $consumable->numRemaining()  }} out of {{ $consumable->qty}}  {{ trans('admin/consumables/general.remaining') }})</h2>
```
Another spot where we're saying "out of"

veenone replied:
noted

uberbrady commented on resources/views/partials/bootstrap-table.blade.php:
```diff
@@ -645,6 +645,11 @@ function cleanFloat(number) {
 
+    
+    function totalFormatter(data) {
+        return 'Total : ' + data.length;
```
This probably needs internationalization too. You can break into Blade syntax here if you need.

veenone replied:
noted, to be fixed

uberbrady commented on resources/views/users/view.blade.php:
```diff
@@ -705,6 +705,7 @@ class="table table-striped snipe-table table-hover"
         </div><!-- /accessories-tab -->
 
         <div class="tab-pane" id="consumables">
+          <h2>Consumable Checkout History</h2>
```
Probably needs internationalization too.

veenone replied:
ah I missed this one.. will do

uberbrady commented on resources/views/users/view.blade.php:
```diff
@@ -726,19 +727,21 @@ class="table table-striped snipe-table table-hover"
                   <td>{!! $consumable->present()->nameUrl() !!}</td>
                   <td>
-                    {!! Helper::formatCurrencyOutput($consumable->purchase_cost) !!}
+                    {!! $consumable->pivot->totalnum !!}
```
This replaces the purchase cost with the pivot's totalnum, and that's probably not what we want to do.

veenone replied:
after looking around other issues on the use case of this field, you're right.
I will update it by adding more column for it

uberbrady commented on resources/views/users/view.blade.php:
```diff
@@ -726,19 +727,21 @@ class="table table-striped snipe-table table-hover"
+                    {!! $consumable->pivot->totalnum !!}
                   </td>
-                  <td>{{ $consumable->created_at }}</td>
+                  <td>{{ $consumable->pivot->created_at }}</td>
```
I think this is the right thing to do here actually - we're looking at a user's list of consumables, we don't care when the *consumable* was `created_at`, we care about when the actual consumable checkout occurred (which is, in fact, the Pivot's `created_at`). That seems pretty fair.

veenone replied:
thanks, so this one can be resolved then?

uberbrady reviewed (changes requested):
Thanks for tackling this for us! If we can get a few fixes in I'm hoping that we can get this merged in. I think the general approach of having checkout notes and quantity in the pivot table definitely makes sense.

I think some careful renaming could be really useful here. We are calling a new value '$total` and `$totalnum` but it seems like that's just a checkout quantity. Could we call it something like `$checkout_qty` or something else like that instead? That'd make it a little clearer when we're reading this years and years and years in the future...

On some of the blades I'm seeing that you're re-purposing some of the visual space that we are already using; instead try and just be additive here (except for that one space where you pretty clearly found a bug; keep that fix in there :) )

As I mentioned in the review, you also included a lot of language files - which I definitely appreciate! But unfortunately, those changes will just get blown out when we next regenerate our data from CrowdIn, our internationalization platform. You'll just want to add strings for `en` and then once that gets imported into CrowdIn, we'd definitely love to have your translations for the other languages you know (which seems like a lot! Very cool :) )

I *do* love the "default 1" trick for the `totalnum` - that's a really clever way to get 'free' backwards-compatibility. Very smart.

If you can make the changes I'm requesting I'd love to get this merged. Thanks again for submitting it to us!

veenone committed:
code adjustments based on pull request review

veenone commented:
Files have been updated according to the review. there's still 1 comment not yet put in resolved since I think I have a little bit different understanding. but this is easily resolved actually

veenone committed:
- forgot to simplify duplicate functions based on review
- revert back route on back button
- update mail template content to include consumable notes

veenone commented:
everything should be according to review feedback now

veenone commented:
@uberbrady , can you review may latest changes?

snipe commented:
We're still reviewing the UX for this and how it relates to recent changes. If it gets merged (or closed), you'll be notified because you're following this thread.

snipe commented:
Blergh - any chance you can resolve the conflicts? (I promise I won't take as long to review this time.)

veenone commented:
Just saw your comment, sorry I also missed this action actually.
It's done now

veenone commented:
I have resolved new conflicts 3 times so far. any chances you will review soon @snipe ?

snipe commented:
This looks like it needs rebasing - I don't think you actually changed 36 files?

mikeroq commented:
Shouldn't the field names on the pivot table be more like `total_num` and `checkout_note`? Or maybe named differently like `qty` and `note`?

At least if there is multiple words it should be in `snake_case` format. 

I'm not 100% what code syle this project follows but that's "standard" Laravel naming convention.

snipe commented:
We typically use `first_last` if it's two words

mikeroq commented:
The note upon checkout is no longer stored in the action log, thus removing it from the activity report. Seems like that should still be logged?

veenone commented:
> This looks like it needs rebasing - I don't think you actually changed 36 files?

I just checked that I made changes on these files, mostly on some resource files for some languages (pardon my ignorance as this was my first PR and didn't know that I only need to update the english one :( )

additionally, I made some adjustments on the email template content as well to include the total number in the notification email

veenone commented:
on the snake case format for some variables, should I make changes on them?  @mikeroq, can you point out on which part of pivot for me to update? and also for the note for action log one, can you point this one out as well ? I will include them

mikeroq commented:
I'm just an observer ;) interested in your feature.

I just try to always name columns with multiple words like `assigned_to` instead of `assignedto`. Just follows best practices and makes for better readability. Same with variables/methods just depends on the code style. `$assignedTo or $assigned_to` vs `$assignedto`

veenone committed:
include checkoutnote in activity log

veenone committed:
refactor 2 main variables to have snake_case format
- totalnum --> checkout_qty
- checkoutnote --> checkout_note

veenone commented:
ok, I have made the changes 😃 
- Enable checkout note on the activity log
![image](https://user-images.githubusercontent.com/3839381/177817735-2b2b6e92-4876-4292-a955-61bc36ecb7de.png)
- Renaming improper variable with snake_case as suggested
  - totalnum --> checkout_qty
  - checkoutnote --> checkout_note

mikeroq commented:
Ok but now the note gets stored in two places, action log, and the pivot table. So I guess the real question is should that be duplicated or just stored in the action log.

Also not sure what the purpose of the checkout notes box is since it's serving the same purpose as the existing notes box (which is the checkout note anyways). Also did you check that components checkout controller?

veenone commented:
Actually why I duplicated the notes box is to differentiate between the notes; which is in my understanding to contain the description / global information of the consumable, and then checkout_notes, which contains comment on each time admin checks out the consumable.

I'm checking the duplication now, but is there any concerns having the same information both in pivot & action log? and how do you quick check this pivot table?

veenone committed:
remove duplication on checkoutcontroller

mikeroq commented:
This is the current checkout page ( I'm on mobile)
![image](https://user-images.githubusercontent.com/55674383/177830619-50522c94-e5c1-4820-849b-f42ba55e4bc4.png)
![image](https://user-images.githubusercontent.com/55674383/177830771-77d5e666-344d-4dd4-8960-e510752f81bf.png)

veenone commented:
for mobile view, it uses different blade file? there's no quantity in your screenshot.
I think I need to look for it first..

snipe commented:
We do not have different blades for desktop v mobile

mikeroq commented:
> for mobile view, it uses different blade file? there's no quantity in your screenshot. I think I need to look for it first..

Those are screens from how it works currently, not your branch. I was just showing what the existing notes box does.

veenone commented:
Ah okay, because I checked on my mobile my branch has shown the same layout already. So here I still keep the original notes for read only purpose as this will be only updated using edit and checkout note will be the one that always updateable on every checkout. They are stored on different table
![Screenshot_20220708-211638.jpg](https://user-images.githubusercontent.com/3839381/178124221-d2bcdefb-fa31-4d71-a4a9-be0bf3cd4248.jpg)

Here notes on the activity log is displaying the checkout notes now

![Screenshot_20220708-211453.jpg](https://user-images.githubusercontent.com/3839381/178124357-c775a35b-e8b1-45b3-a60d-38d269df8ebd.jpg)

veenone commented:
Regarding rebasing, should I rebase this PR to the last commit of develop branch @snipe ?

snipe committed:
Merge pull request #11824 from snipe/fixes/removed_assetlog_from_custom_report

Fixed custom asset report timeout - removed assetlog

snipe committed:
Merge remote-tracking branch 'origin/develop'

snipe committed:
Added Applecare lookup link if manufacturer is apple

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11825 from snipe/features/added_applecare_lookup_link

Added Applecare lookup link if manufacturer is apple

snipe committed:
Merge remote-tracking branch 'origin/develop'

snipe committed:
Merge pull request #11771 from rnelsonee/PAT_test

Fixed #11769: PAT Test hardcoded text

snipe committed:
Add @rnelsonee as a contributor

inietov committed:
Add possible values for custom fields of type checkbox

inietov committed:
Removes format when changing to custom fields elements that doesn't need it

inietov committed:
Save and show the default values properly on custom fields of type checkbox

snipe committed:
Merge pull request #11830 from inietov/fixes/default_values_for_checkboxes_unaccessible

Fixed #9801 #11199 Default Values for Checkbox Field Types Unaccessible

snipe committed:
Merge remote-tracking branch 'origin/develop'

uberbrady committed:
try/catch wrap notifications on checkin/checkout

snipe committed:
Fixed 500 when depreciation is set but no purchase_date

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Ignore warning icon if no purchase date

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11835 from snipe/fixes/depreciation_when_no_purchase_date

Fixes 500 in depreciation on asset view when no purchase date

snipe committed:
Better handle missing files

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Remove DL button if file doesn’t exist

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11836 from snipe/fixes/500_if_image_is_missing

Fixed 500 error if uploaded file is missing from asset view

snipe committed:
Merge remote-tracking branch 'origin/develop'

snipe committed:
Downgraded BS Tables

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11837 from snipe/fixes/disappearing_checkboxes

Downgraded BS Tables

snipe committed:
Bumped hash

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
# Conflicts:

Signed-off-by: snipe <snipe@snipe.net>
#	public/css/dist/all.css
#	public/css/dist/bootstrap-table.css
#	public/js/dist/bootstrap-table.js
#	public/mix-manifest.json

snipe committed:
Larger checkbox in lists

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11838 from snipe/features/larger_checkboxes_in_lists

Larger checkbox in lists

snipe committed:
Merge remote-tracking branch 'origin/develop'

Signed-off-by: snipe <snipe@snipe.net>

# Conflicts:
#	public/css/build/app.css
#	public/css/build/overrides.css
#	public/css/dist/all.css
#	public/mix-manifest.json

snipe committed:
Check for licenses.files permissions

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Fixed license policy method

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11842 from snipe/security/license_file_access

Check for `licenses.files` permissions

snipe committed:
Merge remote-tracking branch 'origin/develop'

Godmartinz committed:
corrects the color choice for the ldap_test_results

inietov committed:
Adds a check for empty location values on import

Godmartinz committed:
corrects color for query and results of ldap sync test

inietov committed:
Add a link used to clear the purchase date input

inietov committed:
Redirect correctly if not enough license seats exists to be checkout

inietov committed:
Delete debugging symbols (just kidding, it's just a dd() function)

inietov committed:
Delete not necessary conditional

inietov committed:
Adds cursor:pointer to indicate a clickable link

snipe committed:
Merge pull request #11762 from geo-chen/master

correcting default mysql port from 3309 to 3306

sunflowerbofh committed:
Fixed: #11871 Replacing (deprecated) patchwork/utf8 by polyfill/mbstring.

snipe committed:
Merge pull request #11856 from inietov/fixes/purchase_date_not_nullable

Fixed Purchase date "readonly" value makes purchase date unable to be nulled in update form [sc-19540]

snipe committed:
Merge pull request #11857 from inietov/fixes/throw_exception_instead_of_redirect_when_no_available_licenseseats

Fixed #11827 Throw exception when there are no available seats for checkout.

snipe committed:
Merge remote-tracking branch 'origin/develop'

snipe committed:
Use BS datepicker’s built in clear button

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11877 from snipe/features/use_built_in_clear_on_datepicker

Use BS datepicker’s built in clear button

snipe committed:
Merge remote-tracking branch 'origin/develop'

inietov committed:
Delete duplicate fields from the importer dropdown menu

inietov committed:
Delete another duplicate that somehow has survived

inietov committed:
Delete another set of duplicates

snipe committed:
Merge pull request #11878 from inietov/fixes/asset_importer_duplicate_fields

Delete duplicate fields from the importer dropdown menu [sc-19547]

snipe committed:
Upgraded fontawesome

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11882 from snipe/security/upgraded_fontawesome

Upgraded fontawesome

dependabot[bot] committed:
Bump codacy/codacy-analysis-cli-action from 4.1.0 to 4.2.0

Bumps [codacy/codacy-analysis-cli-action](https://github.com/codacy/codacy-analysis-cli-action) from 4.1.0 to 4.2.0.
- [Release notes](https://github.com/codacy/codacy-analysis-cli-action/releases)
- [Commits](https://github.com/codacy/codacy-analysis-cli-action/compare/v4.1.0...v4.2.0)

---
updated-dependencies:
- dependency-name: codacy/codacy-analysis-cli-action
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

snipe committed:
Merge pull request #11870 from snipe/dependabot/github_actions/codacy/codacy-analysis-cli-action-4.2.0

Bump codacy/codacy-analysis-cli-action from 4.1.0 to 4.2.0

snipe committed:
Merge pull request #11854 from inietov/fixes/location_issue_during_import

Fixed #11847 404 error during Import if a checkout type of location is specified and no checkout location is passed

snipe committed:
Merge pull request #11764 from BasO12/master

Added notes to audit mail notification

snipe committed:
Add @BasO12 as a contributor

snipe committed:
Merge remote-tracking branch 'origin/develop'

Signed-off-by: snipe <snipe@snipe.net>

# Conflicts:
#	public/css/dist/all.css
#	public/js/build/app.js
#	public/js/dist/all.js
#	public/mix-manifest.json

snipe committed:
Merge pull request #11853 from Godmartinz/ldap_test_results_color

Fixes the color choice for the ldap_test_results

snipe committed:
Updated skins

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Updated dev assets

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Updated prod assets

Signed-off-by: snipe <snipe@snipe.net>

# Conflicts:
#	public/css/dist/skins/skin-black-dark.css
#	public/css/dist/skins/skin-black-dark.min.css
#	public/css/dist/skins/skin-blue-dark.css
#	public/css/dist/skins/skin-blue-dark.min.css
#	public/css/dist/skins/skin-green-dark.css
#	public/css/dist/skins/skin-green-dark.min.css
#	public/css/dist/skins/skin-orange-dark.css
#	public/css/dist/skins/skin-orange-dark.min.css
#	public/css/dist/skins/skin-purple-dark.css
#	public/css/dist/skins/skin-purple-dark.min.css
#	public/css/dist/skins/skin-red-dark.css
#	public/css/dist/skins/skin-red-dark.min.css
#	public/css/dist/skins/skin-yellow-dark.css
#	public/css/dist/skins/skin-yellow-dark.min.css
#	public/mix-manifest.json

snipe committed:
Dev assets

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11671 from chartjes/ch-fix-tests

Fixed unit test suite that was not running, added working browser test

snipe committed:
Merge pull request #11872 from sunflowerbofh/tmp-replace-patchwork-utf8

Fixed #11871 Replacing (deprecated) patchwork/utf8 by polyfill/mbstring.

snipe committed:
Remove phplint (for now)

Signed-off-by: snipe <snipe@snipe.net>

sunflowerbofh committed:
Fixed: #11871 Replacing (deprecated) patchwork/utf8 by polyfill/mbstring.

snipe committed:
Merge pull request #11886 from sunflowerbofh/tmp-replace-patchwork-utf8

Tmp replace patchwork utf8

snipe committed:
Fixed #11889 - API route for status labels asset listing

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11890 from snipe/fixes/statuslabels_assetlist

Fixed #11889 - API route for status labels asset listing

snipe committed:
Fixed #11521 - switch to using status meta from status label name

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added setting to choose dashboard type

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11893 from snipe/features/switch_dash_pie_to_status_type

Fixed #11521 - Add option to switch to using status meta from status label name

uberbrady committed:
Added counts to Dashboard Pie Chart legends and tooltips

snipe committed:
Merge pull request #11894 from uberbrady/improve_piechart_labels_legends

Added counts to Dashboard Pie Chart legends and tooltips

uberbrady committed:
Modified dashboard tooltips to add percentage as well as count

snipe committed:
Merge pull request #11895 from uberbrady/switch_dashboard_piechart_to_percent

Modified dashboard tooltips to add percentage as well as count

snipe committed:
Fixed depreciation API call

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11896 from snipe/fixes/500_on_depreciation

Fixed depreciation API call

snipe committed:
Use EscapeFormula() in CSV export

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Made a conditional escape according to .env

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added Vautia

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Extra whitespace

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11901 from snipe/security/escape_formats_in_csv

Use `EscapeFormula()` in CSV export

inietov committed:
Filter non-asset items from download unaccepted assets report

snipe committed:
Merge pull request #11906 from inietov/fixes/server_error_downloading_unaccepted_assets

Fixed error 500 when downloading asset acceptance report. [sc-19555]

snipe committed:
First UI overhaul

Signed-off-by: snipe <snipe@snipe.net>

Godmartinz committed:
fixes mobile views and clickability of sidebar menu

Godmartinz committed:
off by a pixel

snipe committed:
Added routes, language

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added nag alert for unaccepted assets

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Phrasing!

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Stylistic refactoring

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11908 from snipe/features/nicer_view_assets_ui_for_regular_users

Overhaul of UI for unprivileged users

snipe committed:
Updated lnaguage strings

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11917 from snipe/updates/updated_language_strings

Updated lnaguage strings

snipe committed:
Bumped version

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Dev to master in version.php

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Updated all contributors

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11916 from Godmartinz/gh11868_sidebar_unclickable

fixes mobile views and clickability of sidebar menu

snipe committed:
Dev assets

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Updated link

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Switched to translations for better autoglossonym support

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11924 from snipe/features/autoglossonym_locales

Added autoglossonyms for locales and translations for countries

snipe committed:
Merge remote-tracking branch 'origin/develop'

inietov committed:
Use a regex (gulp) to admit ',' (comma) and ', '(comma plus space) as customfields checkbox value delimiter

snipe committed:
Merge pull request #11925 from inietov/fixes/checkbox_customfields_comma_issues

Fixed #11710 - Validation of checkbox data with multiple values

snipe committed:
Added `start_date` and `end_date` migration to users table

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Adde language strings

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added date casting and fillable to models

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added API select fields and filters

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Store data for start and end

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added start/end to the users API

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added start/end to edit/view screens

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added clear button to generic datepicker and made it wider

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Make the APU return null instead of blank

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added start/end to sortable columns

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11926 from snipe/features/add_start_and_term_dates_to_users

Added start and end dates to users

snipe committed:
Indented divs per @uberbrady

Signed-off-by: snipe <snipe@snipe.net>

uberbrady committed:
Use the new scim-trace feature from our fork of laravel-scim-server lib

snipe committed:
Merge pull request #11927 from uberbrady/use_improved_scim

Use the new scim-trace feature from our fork of laravel-scim-server lib

snipe committed:
Merge remote-tracking branch 'origin/develop'

eltociear committed:
Fix typo in Dockerfile.fpm-alpine

enviroment -> environment

inietov committed:
Send item variable that the datepicker needs

snipe committed:
Merge pull request #11933 from inietov/fixes/500_creating_new_user_undefined_variable_item

Fixed Error 500 undefined variable item creating a new user.

snipe committed:
Merge remote-tracking branch 'origin/develop'

Godmartinz committed:
removes the margin top adjustment for desktop

snipe committed:
Merge pull request #11937 from Godmartinz/gh11868_sidebar_unclickable

Fixes the margin top adjustment for desktop

nh314 committed:
Correct comparison logic

Comparison operator should be greater than or equal

dependabot[bot] committed:
Bump actions/checkout from 2 to 3.1.0

Bumps [actions/checkout](https://github.com/actions/checkout) from 2 to 3.1.0.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v2...v3.1.0)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

nh314 committed:
Remove required mark

Because it actually not required by model.

Godmartinz committed:
adds shortcuts for listing, creating, and saving

snipe committed:
Updated help text to be more clear

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Additional clarification

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
More text tweaking

Signed-off-by: snipe <snipe@snipe.net>

inietov committed:
Add precision parameter to the diffForHumans() Carbon function to report

snipe committed:
Merge pull request #11959 from inietov/fixes/depreciation_date

Fixed Depreciation date diff readable precision

snipe committed:
Sets nullable attribute on validation

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11952 from snipe/features/clearer_activated_helptext

Added clearer LDAP activated helptext

snipe committed:
Merge pull request #11960 from snipe/fixes/better_validation_for_ldap_filter

Sets nullable attribute on validation

snipe committed:
Merge pull request #11942 from nh314/develop

Correct comparison logic

snipe committed:
Merge pull request #11949 from nh314/Remove-required-mark

Remove required mark

snipe committed:
Added @nh314 as a contributor

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11944 from snipe/dependabot/github_actions/actions/checkout-3.1.0

Bump actions/checkout from 2 to 3.1.0

snipe committed:
Set branch back to develop

I hads to switch branches because the automated GH action script always targets masterb

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Use admin ID for audit log notification

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11964 from snipe/fixes/audit_admin_presenting

Use admin ID for audit log notification

snipe committed:
Updated dev assets

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Removed extraneous files

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Updated strings

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Remove unused files

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added strings

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Updated strings

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Updated strings

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11971 from snipe/localization/new_translations

Added new translations

snipe committed:
Bumped version

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11834 from uberbrady/try_catch_notifications

try/catch wrap notifications on checkin/checkout

Godmartinz committed:
gets the proper amount of months and fixes typo of the word depreciation

Godmartinz committed:
adds proper spacing

snipe committed:
Merge pull request #11986 from Godmartinz/FD31357_depreciation

Fixes the calc of months and fixes typo of the word depreciation

snipe committed:
Fixed incorrect icons for mobile view on location tabs

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Fixed user icon for tab

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11988 from snipe/fixes/wrong_icons_for_mobile_location_view

Fixed wrong icons for mobile view for locations listing

snipe committed:
Fixed kit parse error on model edit

Signed-off-by: snipe <snipe@snipe.net>

uberbrady committed:
Try to better handle LDAP Manager settings that don't line up with what we expect

uberbrady committed:
Remove incorrect comment.

uberbrady committed:
Erroneous line-break removal - re-add

snipe committed:
Merge pull request #11995 from uberbrady/ldap_manager_fix

Fixed: Okta LDAP Manager Attribute - Try to better-handle LDAP Manager settings

uberbrady committed:
Caches manager lookups in LDAP for performance boost

snipe committed:
Merge pull request #12006 from uberbrady/ldap_manager_cache

Added cache manager lookups in LDAP for performance boost

snipe committed:
Added default tab to locations

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Updated JS to better handle input IDs

Signed-off-by: snipe <snipe@snipe.net>

ghost commented:
@snipe @uberbrady @mikeroq Is there anything we can do to help land this feature in Snipe-IT?

snipe committed:
Merge pull request #12008 from snipe/features/add_default_loc_to_location_view

Added default tab to locations

snipe committed:
Changed button icon for importer

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #12015 from snipe/fixes/revert_processing_button

Changed button icon for importer

snipe committed:
Fixed typo

Signed-off-by: snipe <snipe@snipe.net>

StarlessNights committed:
Added php memcached extension to the docker image.

StarlessNights committed:
Added php memcached extension to docker Alpine image.

uberbrady committed:
Add externalId support to SCIM integration

uberbrady committed:
Set scim_externalid to nullable, default null

snipe committed:
Merge pull request #12030 from uberbrady/scim_externalid_support

Added externalId support to SCIM integration

snipe committed:
Merge pull request #12028 from StarlessNights/feature/add-memcached-php-extensions-to-docker-images

Fixed #12027: Add memcached php extensions to docker images

snipe committed:
Fixed all-contributors type

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #11928 from eltociear/patch-1

Fix typo in Dockerfile.fpm-alpine

snipe committed:
Add @eltociear as a contributor

snipe committed:
Merge pull request #11951 from Godmartinz/AccessKey_shortcuts

adds shortcuts for listing, creating, and saving

snipe committed:
Cleaned up and added comments

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added Assets Assigned to tab

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #12035 from snipe/features/added_new_tab_to_locations

Features/added new tab to locations

snipe committed:
Null expected_checkin on user bulk checkin and delete

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #12038 from snipe/fixes/clear_expected_checkin_on_checkin_action

Added - set `expected_checkin` to `null` on user bulk checkin and delete

snipe committed:
Renamed column back to assets_count for legacy

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #12040 from snipe/fixes/locations_asset_count

Renamed column back to assets_count for legacy

snipe committed:
Bumped hash

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Reflect color in sidenav if one is given

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Removed extra spaces

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #12041 from snipe/features/reflect_status_label_color_in_sidenav

Fixed #12009 - Added ability to reflect color in sidenav if one is given

snipe committed:
Corrected string for invalid current password

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #12042 from snipe/fixes/added_missing_hashed_password_string

Corrected string for invalid current password

snipe committed:
Merge pull request #11766 from snipe/fixes/wtf_method_names_derp

Fixed (?) method name?

uberbrady commented:
@whalehub I'm re-reviewing the PR now. Regarding rebasing @veenone - yes, it would need to be rebased against the latest develop. I'm hoping to follow-up with a more thorough review shortly.

uberbrady commented on app/Events/CheckoutableCheckedOut.php:
```diff
@@ -14,17 +14,25 @@ class CheckoutableCheckedOut
      * @return void
      */
-    public function __construct($checkoutable, $checkedOutTo, User $checkedOutBy, $note)
+    public function __construct($checkoutable, $checkedOutTo, User $checkedOutBy, $note, $checkout_qty=null, $checkout_note=null)
```
I wonder if this `$checkout_qty` default should be `1`?

veenone replied:
fixed

uberbrady commented on database/migrations/2013_11_25_013244_create_licenses_table.php:
```diff
@@ -0,0 +1,43 @@
+<?php
```
This one doesn't seem quite this file ought to be here?

veenone replied:
removed from my branch

uberbrady commented on database/migrations/2022_03_18_104400_add_total_to_consumables_users.php:
```diff
@@ -0,0 +1,38 @@
+<?php
```
This one either

veenone replied:
this one is the migration script for my feature, so I keep it here

uberbrady reviewed (changes requested):
It's still a bit hard to parse what here is changing and what isn't - if I could get a fresh rebase I can take another look. I still like the idea; and I think we should do it. I think the architectural choices all seem solid. Thank you!

snipe committed:
Check for valid category

Signed-off-by: snipe <snipe@snipe.net>

uberbrady committed:
Needed to fix the way we write externalId from SCIM

veenone commented:
I'm sorry I haven't been able to check the project these past weeks.
I will try to rebase it then response to the request :)

snipe committed:
Added consumables upload routes

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added consumable uploads gitignore

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added consumable methods for uploads

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added consumables permissions

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Updated consumables UI to allow uploads and display files

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Skip storage:: facade on missing images

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Fixed some layout issues

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #12057 from snipe/features/add_uploads_to_components

Added uploads to components

snipe committed:
Updated permissions

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added helper method for uploads

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added component routes for uploads

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added ComponentFilesController

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Fixed some translations

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added some generic file translations

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Updated component blade

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Fixed component file route controller names

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Added gitignore

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #12058 from snipe/features/added_uploads_to_components

Added uploads to components

snipe committed:
Use the generic file uploads strings

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #12059 from snipe/fixes/consumables_files_translations

Use the generic file uploads strings

snipe committed:
Accessory file upload

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #12060 from snipe/features/added_accessories_file_uploads

Added accessories file uploads

snipe committed:
Merge pull request #12049 from uberbrady/improve_scim_externalid_support

Fixed writes to the `externalId` from SCIM

snipe committed:
Fixed image path to show preview

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Disallow uploads if app is locked

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Use new upload string

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #12066 from snipe/fixes/small_uploads_fixes

Fixed - small uploads fixes

snipe committed:
Fixed model file image path/variable

Signed-off-by: snipe <snipe@snipe.net>

snipe committed:
Merge pull request #12067 from snipe/fixes/modelfile_image_path

Fixed model file image path/variable

snipe committed:
Bumped version

Signed-off-by: snipe <snipe@snipe.net>

veenone committed:
Update based on input

- remove unrelated migration script
- update default checkout qty to 1

veenone committed:
Update based on input

- remove unrelated migration script
- update default checkout qty to 1

veenone committed:
Merge branch 'feat/multiple_consumables_checkout' of https://github.com/veenone/snipe-it into feat/multiple_consumables_checkout

veenone commented:
hi @uberbrady , I have rebased my branch.
I hope this rebase is properly done... looking forward for your review

veenone commented:
Happy new year @snipe @uberbrady!
I hope this year one of my PRs can be reviewed and integrated .. 🤞

ghanendaghestani commented:
Any idea when this feature/fix will be available. It is impacting our origination as well. Thanks

rbarreiros commented:
Any update on this ? This would be a massive improvement!

jacojazz commented:
Can't wait for this to be added!

jayavman commented:
@snipe any chance of this getting merged soon?
